### PR TITLE
Improve leadership change signaling

### PR DESCRIPTION
**Pull request by jnmoyne**
Because leadership status can sometimes change very quickly and raft.updateLeadChange() can 'drop' some lead change signals, especially with the leadc channel having a capacity of one, which means that sometimes stream.setLeader(isLeader bool) gets called twice in a row with the same isLeader value.

This could lead to in some cases the sourcing to 'stall' (not all messages gets sourced) or 'duplicate messages' (more messages than expected get sourced).

Changes are increasing the leadc channel's size to 32.

The effects of the change can be demonstrated by running the TestStreamSourcingScalingSourcingManyBenchmark load test a large number of times (e.g. 100 times): without these changes the test flaps and fails between 10 and 20% of the time (most of the time 'stalling' but also sometimes with 'too many'). With the changes over around a hundred runs not once did the test fail with 'stalling' and only once did it fail with 'too many' messages (so there's still a bug/corner case that could result in 'more than expected' messages to be sourced, although it seems to be a rare occurence (i.e. 1% of the time)).

This is a replacement for #5496

### CODE REVIEW

File: server/raft.go [hunk 1]
```diff
@@ -387,7 +387,7 @@ func (s *Server) startRaftNode(accName string, cfg *RaftConfig, labels pprofLabe
 		apply:    newIPQueue[*CommittedEntry](s, qpfx+"committedEntry"),
 		stepdown: newIPQueue[string](s, qpfx+"stepdown"),
 		accName:  accName,
-		leadc:    make(chan bool, 1),
+		leadc:    make(chan bool, 32),
```
Comment (neilalexander) [inline server/raft.go:390]: Wonder if it is better for this to be an `*ipQueue[bool]` instead? I'd like to hope that we don't switch ever leadership state that quickly 32 times, but if it's important that we never lose that transition, IP queues don't have backpressure like plain channels do.

File: server/jetstream_sourcing_scaling_test.go [hunk 2]
```diff
@@ -14,259 +14,126 @@
 package server
 
 import (
-	"encoding/json"
 	"fmt"
+	"github.com/nats-io/nats.go"
```
Comment (derekcollison) [inline server/jetstream_sourcing_scaling_test.go:18]: I know a PITA but let's put proper tab indentation on these configs.

File: server/stream.go [hunk 3]
```diff
@@ -839,6 +839,28 @@ func (mset *stream) isLeader() bool {
 	return true
 }
 
+// Lock should be held
+func (mset *stream) stepDown() {
+	// cancel timer to create the source consumers if not fired yet
```
Comment (derekcollison) [inline server/stream.go]: Start comments with caps.

File: server/stream.go [hunk 4]
```diff
@@ -839,6 +839,28 @@ func (mset *stream) isLeader() bool {
...
+	if mset.sourcesConsumerSetup != nil {
+		mset.sourcesConsumerSetup.Stop()
+		mset.sourcesConsumerSetup = nil
+	} else { // otherwise cancel the source consumers
```
Comment (derekcollison) [inline server/stream.go]: Should we qualify if needed here?

File: server/stream.go [hunk 5]
```diff
@@ -876,14 +854,23 @@ func (mset *stream) setLeader(isLeader bool) error {
...
 			mset.mu.Unlock()
 			return err
 		}
 	} else {
-		mset.stepDown()
+		// cancel timer to create the source consumers if not fired yet
+		if mset.sourcesConsumerSetup != nil {
+			mset.sourcesConsumerSetup.Stop()
+			mset.sourcesConsumerSetup = nil
+		}
```
Comment (derekcollison) [inline server/stream.go:867]: This else clause got dropped from above `mset.stopSourceConsumers()`

File: server/raft.go (hunk 1 above)
Comment (derekcollison) [inline server/raft.go:390]: Also did we want to change the second conditional in placing values on this like we discussed?

**Review by derekcollison — CHANGES_REQUESTED** [review]

File: server/stream.go (hunk 4 above)
Comment (jnmoyne) [reply server/stream.go]: Technically the stopSourceConsumers() could be always executed rather than only if mset.sourcesConsumerSetpup != nil.

But it is now removed along with all the other cleanup I had added initially.

File: server/stream.go (hunk 5 above)
Comment (jnmoyne) [reply server/stream.go:867]: The else clause was part of the cleanup that I had initially added and that you asked me to remove

File: server/stream.go (hunk 3 above)
Comment (jnmoyne) [reply server/stream.go]: Now removed along with all the other cleanup stuff I had added initially

File: server/jetstream_sourcing_scaling_test.go (hunk 2 above)
Comment (jnmoyne) [reply server/jetstream_sourcing_scaling_test.go:18]: ? it is with a tab indentation for me. 
(I use gofmt to reformat)

File: server/stream.go (hunk 5 above)
Comment (derekcollison) [reply server/stream.go:867]: Let's add that one in, but put conditional so code reads a bit cleaner, so `if ... {`

File: server/raft.go (hunk 1 above)
Comment (jnmoyne) [reply server/raft.go:390]: In my experimentation with the `TestStreamSourcingScalingSourcingManyBenchmark` test it only flipped quickly once in a row. But it could certainly be possible that in other scenarios it could flip quickly more than once, but 32 is a pretty high number so (at least in my testing) it doesn't ever get that high and the limit (where signals start getting dropped) is never reached.

Note that in the previous version (#5496) I had not touched the size of the channel and instead made `setLeader()` remember what state it is in and force a cleanup and re-setup in case it was already the leader and it gets a setLeader(true) (which would happen when a signal of false was followed a fraction of a second later with a signal of true - and therefore the signal of false was dropped by the channel because of capacity of 1). But this is not there anymore in this version of the PR.

File: server/jetstream_sourcing_scaling_test.go (hunk 2 above)
Comment (derekcollison) [reply server/jetstream_sourcing_scaling_test.go:18]: Manually put tabs in front of lines on configs, like we do in other tests etc. So reads easier, parser will handle.

Comment (jnmoyne) [reply server/jetstream_sourcing_scaling_test.go:18]: Sorry mis-understood which part of the code the comment was for, added tabs to the configs.

File: server/stream.go (hunk 4 above)
Comment (jnmoyne) [reply server/stream.go]: Added back

File: server/stream.go (hunk 5 above)
Comment (jnmoyne) [reply server/stream.go:870]: ok done.

**Review by derekcollison — APPROVED** [review]
LGTM
